### ISSUE TITLE
Behaviour driven developmen; to be, or not to be :)

### ISSUE THREAD

**Issue by escanda** [gopacifia/DEEP#36]
As I am digging into the model of valuenetwork, I'm finding myself a bit lost about all interactions that happen at a given moment an action occurs, such as creating new entities, changing other's states and so on.

Although I couldn't have the chance yet to implement BDD in any commercial project, we were planning to do so on a hospital warehouse inventory management tool, but we didn't finally get up to that point.

I noticed Python has Behave's framework which should do the trick. You can read more about it at http://pythonhosted.org/behave/philosophy.html

The main arguments imho to have these tests are these:
- *it helps newcomers* such as myself understand model interactions at a glance of user stories.
- it helps non technical people to create test stories to be implemented later on by coders.
- more tests prevent regressions.
- it helps organize tests in more comprehensible units than unit tests for large fetched stories.
(plus all the reasons mentioned in the link above in the paragraph "BDD practices")

In this matter I think @fosterlynn, @bhaugen and @numapanuma could be the most indicated persons to write the initial user stories. Specially @fosterlynn and @bhaugen since the model is known to them as the palm of their hand :) @numapanuma could write user stories of future features I think, once the core ones are ok.

This is just a question and proposal, any criticism is welcome :)

Specially the criticism I might say as I understand we are on a tight schedule and resources. This may be kind of optional.

So, what do you think? Do you have any good or bad experience with BDD in your projects?

**Comment by bhaugen** [comment]
I love the idea of BDD, and was influenced by it somewhat when I did this set of tests:
https://github.com/gopacifia/DEEP/blob/master/valuenetwork/valueaccounting/tests/test_value_equations.py

I wrote those tests before I made a lot of changes to the value equation code, which is probably the most difficult code in the system, and we need it to be correct. So it was also an example of test-driven development, which I sometimes do, but not always, and not strictly. 

(I am mostly a get-it-done kind of programmer. But in this case, I was afraid the code would get out of control, and having the tests upfront takes away a lot of the fear.)

Had some friends in the Value Flows project that were using some combination of BDD and Domain-Driven Design (DDD). They were using https://cucumber.io/

You can see the start of it here: https://github.com/valueflows/resourceful

Here's where agent lewis explains his evolving methodology:
https://github.com/valueflows/resourceful/issues/5

As far as I know, they never got done. That's not a warning. And they are helpful people, you could raise an issue on that repo and get answers to questions.

**Comment by fosterlynn** [comment]
I don't have experience with BDD specifically.  So a question and a thought: if we do want to experiment with it, can we do it piecemeal?  That is, as people start to work on a feature, document the user story(s) for just that work?  Might help with the schedule issue.

**Comment by escanda** [comment]
@bhaugen I have in my bookshelf DDD by Evans and I am a big fan for business apps of his ideas. As a matter of fact, I read in some docs you and @fosterlynn based some of your patterns on http://phruby.com/book/index.html which looks really nice although the price tag let me a bit down. I have to explore some other book shops rather than Amazon to see whether they have it cheaper. To me, the feature book, driven by specs sounds a fantastic idea; if I am not mistaken, Behave (Python's cucumber), "speaks" Gherkins which makes it amenable to other existing tools as Cucumber.

@fosterlynn Sounds good :) I agree this is not a priority by itself, but I think it could add a lot of value to more complex issues such as #32 where @numapanuma could write down the spec, I could write the tests, and then implement the code.

**Comment by bhaugen** [comment]
We work in spirals between diving deeply into particular networks and generalizing: summarizing the results and lessons.

We generalize in https://valueflo.ws/ which is an example of domain-driven design. That's what the Resourceful gang was using for their domain model.

**Comment by fosterlynn** [comment]
@escanda pending looking at Pavel's book :), here are links to a bunch of reading on REA, the pattern / domain model that we based NRP/DEEP on, in case that helps you.  https://github.com/valnet/valuenetwork/wiki/Links

Also other reading here: https://github.com/valnet/valuenetwork/wiki  (The wiki from the original code didn't make it through the forking process.)  Maybe check Core, Concepts, ?.  And if you really are a glutton for punishment, check the Tutorials, https://speakerdeck.com/mikorizal.

**Comment by escanda** [comment]
@bhaugen makes sense. A bit of an unrelated anecdote, while I was working on feature of a warehouse mgmt system, the ERP people were working on traceability, in order to track products across warehouses, shops and so on; REA with its agents captures well and generically enough these kind of events. I have to dig yet, how REA deals with traditional accounting. For instance, at the end of the year, how to fill taxes :) I can make myself an idea running a few queries, but taxes and so on are a pita.

@fosterlynn I didn't know of the Links page in the wiki, really interesting stuff. Thanks :) I read most of the tutorials, but I will reread them once I setup the instance we talked about in telegram in order to track development of DEEP in there to a finer detail: logging work hours, asking for assistance, information exchanges, etc. So I can have a better picture as a user of the system. Perhaps we can mix kudos and time to see how it works out. Specially in reference to #1 (not sure if it makes sense to redistribute kudos though :)

**Comment by escanda** [comment]
BTW, I made an entity diagram using:
`python manage.py graph_models --pygraphviz --all-applications -g > model.dot`
`dot -T png model.dot -o model.png`

Result is attached.
![model](https://cloud.githubusercontent.com/assets/13382418/22863584/54cd6b5a-f143-11e6-83e0-862e60da4fbf.png)

**Comment by bhaugen** [comment]
Wow! @escanda you are cookin! (But that diagram is so big I can't follow it myself. Not your fault, it's the models. We should clean them up and just put the core into models.py and the other stuff in a related module.)

> the ERP people were working on traceability, in order to track products across warehouses, shops and so on; REA with its agents captures well and generically enough these kind of events. 

Yes. We worked on a food network system based on REA that did strict lot tracking and tracing from source to end use and back. One of the uses was for tracing mad cow disease back to the source animal.

> I have to dig yet, how REA deals with traditional accounting. 

Here are a couple of discussions of that topic:
https://github.com/valueflows/valueflows/issues/141
https://github.com/valueflows/valueflows/issues/142

We've done some of it, but other people in the REA community have done a lot more.

> For instance, at the end of the year, how to fill taxes

We've never done that automatically from any of our REA-based software. But other people have used the information to file taxes.

**Comment by numapanuma** [comment]
I think it's a great idea to implement BDD, for all the reasons you mentionned @escanda. Any action that would help new comers get into the project more easily would be very important. If @fosterlynn and @bhaugen think I can help writing stories for the BDD I would happy to contribute!

**Comment by bhaugen** [comment]
@numapanuma 
> If @fosterlynn and @bhaugen think I can help writing stories for the BDD I would happy to contribute!

Yes, you would be a good source of BDD stories.

**Comment by fosterlynn** [comment]
@numapanuma maybe you can even start, since you reviewed the software so thoroughly awhile back. :)  With @escanda 's help to use the tool of his choice, and figure out how to think about it.  (We are pretty busy this week, and part way into next week (we're going to a MAN summit and need to get some ValueFlows things ready for that too.)

And then you both can manage the possibly competing priorities.  You could start with the areas that need the highest priority work.

Then I'd be happy to pitch in and correct, add to, etc., based on knowing the code.
